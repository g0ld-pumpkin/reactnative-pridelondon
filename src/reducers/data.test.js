// @flow
import {
  generateCMSEvent,
  generateCMSParadeGroup,
  generateCMSSponsor,
  generateCMSAmenity,
  sampleOne
} from "../data/__test-data";
import reducer from "./data";

describe("Events reducer", () => {
  it("initialises with default state", () => {
    // $FlowFixMe
    const state = reducer(undefined, {});

    expect(state).toMatchSnapshot();
  });

  it("sets loading flag for REQUEST_CMS_DATA action", () => {
    const initialState = {
      events: [],
      featuredEvents: [],
      headerBanners: [],
      images: {},
      paradeGroups: [],
      performances: {},
      sponsors: [],
      amenities: [],
      loading: false,
      refreshing: false,
      noDataReceived: false
    };
    const state = reducer(initialState, { type: "REQUEST_CMS_DATA" });

    expect(state.loading).toBe(true);
    expect(state.refreshing).toBe(false);
  });

  it("sets refreshing flag for REQUEST_UPDATE_CMS_DATA action", () => {
    const initialState = {
      events: [],
      featuredEvents: [],
      headerBanners: [],
      images: {},
      paradeGroups: [],
      performances: {},
      sponsors: [],
      amenities: [],
      loading: false,
      refreshing: false,
      noDataReceived: false
    };
    const state = reducer(initialState, { type: "REQUEST_UPDATE_CMS_DATA" });

    expect(state.loading).toBe(false);
    expect(state.refreshing).toBe(true);
  });

  it("sets loading and refreshing to false for NO_DATA_RECEIVED action", () => {
    const initialState = {
      events: [],
      featuredEvents: [],
      headerBanners: [],
      images: {},
      paradeGroups: [],
      performances: {},
      sponsors: [],
      amenities: [],
      loading: true,
      refreshing: true,
      noDataReceived: false
    };
    const state = reducer(initialState, { type: "NO_DATA_RECEIVED" });

    expect(state.loading).toBe(false);
    expect(state.refreshing).toBe(false);
    expect(state.noDataReceived).toBe(true);
  });

  describe("RECEIVE_CMS_DATA action", () => {
    it("decodes events", () => {
      const initialState = {
        events: [],
        featuredEvents: [],
        headerBanners: [],
        images: {},
        paradeGroups: [],
        performances: {},
        sponsors: [],
        amenities: [],
        loading: true,
        refreshing: false,
        noDataRecived: true
      };

      const newCmsData = {
        entries: [sampleOne(generateCMSEvent, { seed: 1345 })],
        assets: [],
        syncToken: "abc",
        updated: true
      };

      // $FlowFixMe
      const state = reducer(initialState, {
        type: "RECEIVE_CMS_DATA",
        data: newCmsData
      });

      expect(state.events).toMatchSnapshot();
      expect(state.noDataReceived).toBe(false);
    });

    it("expands recurring events", () => {
      const initialState = {
        events: [],
        featuredEvents: [],
        headerBanners: [],
        images: {},
        paradeGroups: [],
        performances: {},
        sponsors: [],
        amenities: [],
        loading: true,
        refreshing: false,
        noDataRecived: false
      };

      const event = sampleOne(generateCMSEvent, { seed: 3353 });
      if (event && event.fields && typeof event.fields === "object") {
        event.fields.startTime = { "en-GB": "2018-08-02T00:00+00:00" };
        event.fields.endTime = { "en-GB": "2018-08-02T03:00+00:00" };
        event.fields.recurrenceDates = {
          "en-GB": ["03/08/2018", "04/08/2018"]
        };
      }

      const newCmsData = {
        entries: [event],
        assets: [{ id: "1" }],
        syncToken: "abc",
        updated: true
      };

      // $FlowFixMe
      const state = reducer(initialState, {
        type: "RECEIVE_CMS_DATA",
        data: newCmsData
      });

      expect(state.events).toMatchSnapshot();
    });

    it("decodes featuredEvents", () => {
      const initialState = {
        events: [],
        featuredEvents: [],
        headerBanners: [],
        images: {},
        paradeGroups: [],
        performances: {},
        sponsors: [],
        amenities: [],
        loading: true,
        refreshing: false
      };

      const newCmsData = {
        entries: [
          {
            fields: {
              title: { "en-GB": "title" },
              events: { "en-GB": [] }
            },
            sys: {
              id: "3O3SZPgYl2MUEWu2MoK2oi",
              contentType: {
                sys: {
                  id: "featuredEvents"
                }
              },
              revision: 1
            }
          }
        ],
        assets: [],
        syncToken: "abc",
        updated: true
      };

      const expected = [
        {
          id: "3O3SZPgYl2MUEWu2MoK2oi",
          contentType: "featuredEvents",
          revision: 1,
          locale: "en-GB",
          fields: {
            title: "title",
            events: []
          }
        }
      ];

      // $FlowFixMe
      const state = reducer(initialState, {
        type: "RECEIVE_CMS_DATA",
        data: newCmsData
      });

      expect(state.featuredEvents).toEqual(expected);
    });

    it("decodes headerBanners", () => {
      const initialState = {
        events: [],
        featuredEvents: [],
        headerBanners: [],
        images: {},
        paradeGroups: [],
        performances: {},
        sponsors: [],
        amenities: [],
        loading: true,
        refreshing: false
      };

      const newCmsData = {
        entries: [
          {
            fields: {
              heading: { "en-GB": "heading" },
              headingLine2: { "en-GB": "headingLine2" },
              subHeading: { "en-GB": "subHeading" },
              heroImage: { "en-GB": { sys: { id: "2o2SZPgYl2ABCWu2MoK333" } } },
              backgroundColour: { "en-GB": "#333333" }
            },
            sys: {
              id: "3O3SZPgYl2MUEWu2MoK2oi",
              contentType: {
                sys: {
                  id: "headerBanner"
                }
              },
              revision: 1
            }
          }
        ],
        assets: [],
        syncToken: "abc",
        updated: true
      };

      const expected = [
        {
          id: "3O3SZPgYl2MUEWu2MoK2oi",
          contentType: "headerBanner",
          revision: 1,
          locale: "en-GB",
          fields: {
            heading: "heading",
            headingLine2: "headingLine2",
            subHeading: "subHeading",
            heroImage: { sys: { id: "2o2SZPgYl2ABCWu2MoK333" } },
            backgroundColour: "#333333"
          }
        }
      ];

      // $FlowFixMe
      const state = reducer(initialState, {
        type: "RECEIVE_CMS_DATA",
        data: newCmsData
      });

      expect(state.headerBanners).toEqual(expected);
    });

    it("decodes performances", () => {
      const initialState = {
        events: [],
        featuredEvents: [],
        headerBanners: [],
        images: {},
        paradeGroups: [],
        performances: {},
        sponsors: [],
        amenities: [],
        loading: true,
        refreshing: false
      };

      const newCmsData = {
        entries: [],
        assets: [
          {
            fields: {
              file: {
                "en-GB": {
                  url: "//localhost/image.jpg",
                  details: {
                    image: {
                      height: 100,
                      width: 100
                    }
                  }
                }
              }
            },
            sys: {
              id: "3O3SZPgYl2MUEWu2MoK2oi",
              revision: 1
            }
          }
        ],
        syncToken: "abc",
        updated: true
      };

      const expected = {
        "3O3SZPgYl2MUEWu2MoK2oi": {
          id: "3O3SZPgYl2MUEWu2MoK2oi",
          revision: 1,
          uri: "https://localhost/image.jpg",
          width: 100,
          height: 100
        }
      };

      // $FlowFixMe
      const state = reducer(initialState, {
        type: "RECEIVE_CMS_DATA",
        data: newCmsData
      });

      expect(state.images).toEqual(expected);
    });

    it("decodes parade groups", () => {
      const initialState = {
        events: [],
        featuredEvents: [],
        headerBanners: [],
        images: {},
        paradeGroups: [],
        performances: {},
        sponsors: [],
        amenities: [],
        loading: true,
        refreshing: false
      };

      const newCmsData = {
        entries: [sampleOne(generateCMSParadeGroup, { seed: 6534 })],
        assets: [],
        syncToken: "abc",
        updated: true
      };

      // $FlowFixMe
      const state = reducer(initialState, {
        type: "RECEIVE_CMS_DATA",
        data: newCmsData
      });

      // $FlowFixMe
      expect(state.paradeGroups[0].id).toEqual(newCmsData.entries[0].sys.id);
    });

    it("decodes performances", () => {
      const initialState = {
        events: [],
        featuredEvents: [],
        headerBanners: [],
        images: {},
        paradeGroups: [],
        performances: {},
        sponsors: [],
        amenities: [],
        loading: true,
        refreshing: false
      };

      const newCmsData = {
        entries: [
          {
            fields: {
              title: { "en-GB": "title" },
              startTime: { "en-GB": "2018-07-07T12:00:00+01:00" }
            },
            sys: {
              id: "3O3SZPgYl2MUEWu2MoK2oi",
              contentType: {
                sys: {
                  id: "performance"
                }
              },
              revision: 1
            }
          }
        ],
        assets: [],
        syncToken: "abc",
        updated: true
      };

      const expected = {
        "3O3SZPgYl2MUEWu2MoK2oi": {
          id: "3O3SZPgYl2MUEWu2MoK2oi",
          contentType: "performance",
          revision: 1,
          locale: "en-GB",
          fields: {
            title: "title",
            startTime: "2018-07-07T12:00:00+01:00"
          }
        }
      };

      // $FlowFixMe
      const state = reducer(initialState, {
        type: "RECEIVE_CMS_DATA",
        data: newCmsData
      });

      expect(state.performances).toEqual(expected);
    });

    it("decodes sponsors", () => {
      const initialState = {
        events: [],
        featuredEvents: [],
        headerBanners: [],
        images: {},
        paradeGroups: [],
        performances: {},
        sponsors: [],
        amenities: [],
        loading: true,
        refreshing: false
      };

      const newCmsData = {
        entries: [sampleOne(generateCMSSponsor, { seed: 6534 })],
        assets: [],
        syncToken: "abc",
        updated: true
      };

      // $FlowFixMe
      const state = reducer(initialState, {
        type: "RECEIVE_CMS_DATA",
        data: newCmsData
      });

      // $FlowFixMe
      expect(state.sponsors[0].id).toEqual(newCmsData.entries[0].sys.id);
    });

    it("decodes amenities", () => {
      const initialState = {
        events: [],
        featuredEvents: [],
        headerBanners: [],
        images: {},
        paradeGroups: [],
        performances: {},
        sponsors: [],
        amenities: [],
        loading: true,
        refreshing: false
      };

      const newCmsData = {
        entries: [sampleOne(generateCMSAmenity, { seed: 6534 })],
        assets: [],
        syncToken: "abc",
        updated: true
      };

      // $FlowFixMe
      const state = reducer(initialState, {
        type: "RECEIVE_CMS_DATA",
        data: newCmsData
      });

      // $FlowFixMe
      expect(state.amenities[0].id).toEqual(newCmsData.entries[0].sys.id);
    });
  });
});
