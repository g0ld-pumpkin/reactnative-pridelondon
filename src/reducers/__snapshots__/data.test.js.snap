// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Events reducer RECEIVE_CMS_DATA action decodes events 1`] = `
Array [
  Object {
    "contentType": "event",
    "fields": Object {
      "accessibilityDetails": "accessibilityDetails",
      "accessibilityOptions": Array [
        "O84a35poabxKTJm2d5",
      ],
      "addressLine1": "addressLine1",
      "addressLine2": "addressLine2",
      "audience": Array [
        "p7m",
        "75ckc33EMZTM3",
      ],
      "city": "city",
      "email": "email",
      "endTime": "2018-07-07T03:00+00:00",
      "eventCategories": Array [
        "Nightlife",
        "Social and Networking",
        "Community",
        "Health",
        "Sports and Activities",
      ],
      "eventDescription": "eventDescription",
      "eventPriceHigh": 10,
      "eventPriceLow": 0,
      "eventsListPicture": Object {
        "sys": Object {
          "id": "MAj35v",
        },
      },
      "individualEventPicture": Object {
        "sys": Object {
          "id": "385c4r7oSUdL3lRN",
        },
      },
      "location": Object {
        "lat": 0,
        "lon": 10,
      },
      "locationName": "locationName",
      "name": "name",
      "performances": Array [
        Object {
          "sys": Object {
            "id": "T5T",
          },
        },
        Object {
          "sys": Object {
            "id": "dt29BrPh",
          },
        },
      ],
      "phone": "phone",
      "postcode": "postcode",
      "recurrenceDates": Array [
        "10/07/2018",
        "07/07/2018",
        "20/06/2018",
      ],
      "stage": false,
      "startTime": "2018-07-07T00:00+00:00",
      "ticketingUrl": "ticketingUrl",
      "venueDetails": Array [
        "6em2PtRZIhYQiRF44",
        "Z7Z7T6bfuo7qFFz4s3fVA1LYD3lnx",
        "XLrUX5MxwxR9qRvT8t4F3",
        "7LIDGAlkK3Xg",
        "jP6n64T086sxr3sTpxGPLO05J",
      ],
    },
    "id": "r22",
    "locale": "en-GB",
    "revision": 1,
  },
  Object {
    "contentType": "event",
    "fields": Object {
      "accessibilityDetails": "accessibilityDetails",
      "accessibilityOptions": Array [
        "O84a35poabxKTJm2d5",
      ],
      "addressLine1": "addressLine1",
      "addressLine2": "addressLine2",
      "audience": Array [
        "p7m",
        "75ckc33EMZTM3",
      ],
      "city": "city",
      "email": "email",
      "endTime": "2018-07-10T03:00+00:00",
      "eventCategories": Array [
        "Nightlife",
        "Social and Networking",
        "Community",
        "Health",
        "Sports and Activities",
      ],
      "eventDescription": "eventDescription",
      "eventPriceHigh": 10,
      "eventPriceLow": 0,
      "eventsListPicture": Object {
        "sys": Object {
          "id": "MAj35v",
        },
      },
      "individualEventPicture": Object {
        "sys": Object {
          "id": "385c4r7oSUdL3lRN",
        },
      },
      "location": Object {
        "lat": 0,
        "lon": 10,
      },
      "locationName": "locationName",
      "name": "name",
      "performances": Array [
        Object {
          "sys": Object {
            "id": "T5T",
          },
        },
        Object {
          "sys": Object {
            "id": "dt29BrPh",
          },
        },
      ],
      "phone": "phone",
      "postcode": "postcode",
      "recurrenceDates": Array [
        "07/07/2018",
        "10/07/2018",
        "07/07/2018",
        "20/06/2018",
      ],
      "stage": false,
      "startTime": "2018-07-10T00:00+00:00",
      "ticketingUrl": "ticketingUrl",
      "venueDetails": Array [
        "6em2PtRZIhYQiRF44",
        "Z7Z7T6bfuo7qFFz4s3fVA1LYD3lnx",
        "XLrUX5MxwxR9qRvT8t4F3",
        "7LIDGAlkK3Xg",
        "jP6n64T086sxr3sTpxGPLO05J",
      ],
    },
    "id": "r22-recurrence-10/07/2018",
    "locale": "en-GB",
    "revision": 1,
  },
  Object {
    "contentType": "event",
    "fields": Object {
      "accessibilityDetails": "accessibilityDetails",
      "accessibilityOptions": Array [
        "O84a35poabxKTJm2d5",
      ],
      "addressLine1": "addressLine1",
      "addressLine2": "addressLine2",
      "audience": Array [
        "p7m",
        "75ckc33EMZTM3",
      ],
      "city": "city",
      "email": "email",
      "endTime": "2018-06-20T03:00+00:00",
      "eventCategories": Array [
        "Nightlife",
        "Social and Networking",
        "Community",
        "Health",
        "Sports and Activities",
      ],
      "eventDescription": "eventDescription",
      "eventPriceHigh": 10,
      "eventPriceLow": 0,
      "eventsListPicture": Object {
        "sys": Object {
          "id": "MAj35v",
        },
      },
      "individualEventPicture": Object {
        "sys": Object {
          "id": "385c4r7oSUdL3lRN",
        },
      },
      "location": Object {
        "lat": 0,
        "lon": 10,
      },
      "locationName": "locationName",
      "name": "name",
      "performances": Array [
        Object {
          "sys": Object {
            "id": "T5T",
          },
        },
        Object {
          "sys": Object {
            "id": "dt29BrPh",
          },
        },
      ],
      "phone": "phone",
      "postcode": "postcode",
      "recurrenceDates": Array [
        "07/07/2018",
        "10/07/2018",
        "07/07/2018",
        "20/06/2018",
      ],
      "stage": false,
      "startTime": "2018-06-20T00:00+00:00",
      "ticketingUrl": "ticketingUrl",
      "venueDetails": Array [
        "6em2PtRZIhYQiRF44",
        "Z7Z7T6bfuo7qFFz4s3fVA1LYD3lnx",
        "XLrUX5MxwxR9qRvT8t4F3",
        "7LIDGAlkK3Xg",
        "jP6n64T086sxr3sTpxGPLO05J",
      ],
    },
    "id": "r22-recurrence-20/06/2018",
    "locale": "en-GB",
    "revision": 1,
  },
]
`;

exports[`Events reducer RECEIVE_CMS_DATA action expands recurring events 1`] = `
Array [
  Object {
    "contentType": "event",
    "fields": Object {
      "accessibilityDetails": "accessibilityDetails",
      "accessibilityOptions": Array [
        "FM6jre",
        "b8Y2Ioy1qEC",
        "0D98Lcd5il86y1sf4685Ba3s3RbTi4",
      ],
      "addressLine1": "addressLine1",
      "addressLine2": "addressLine2",
      "audience": Array [
        "X2T77u30HfGQ978Ce",
        "IsJqxf058AxsLcfAP7451huX7",
        "",
      ],
      "city": "city",
      "email": "email",
      "endTime": "2018-08-02T03:00+00:00",
      "eventCategories": Array [
        "Exhibition and Tours",
        "Community",
      ],
      "eventDescription": "eventDescription",
      "eventPriceHigh": 10,
      "eventPriceLow": 0,
      "eventsListPicture": Object {
        "sys": Object {
          "id": "7qByW",
        },
      },
      "individualEventPicture": Object {
        "sys": Object {
          "id": "pI1b5E4m4z",
        },
      },
      "location": Object {
        "lat": 0,
        "lon": 10,
      },
      "locationName": "locationName",
      "name": "name",
      "performances": Array [
        Object {
          "sys": Object {
            "id": "97hZrki0698g88ozhhp",
          },
        },
        Object {
          "sys": Object {
            "id": "48V92917",
          },
        },
        Object {
          "sys": Object {
            "id": "u6dtAeCrx2QtSFw7tlYU8Bt2Pgh",
          },
        },
        Object {
          "sys": Object {
            "id": "Qo9LdGIwhTV3VTYF5U5E1g4",
          },
        },
      ],
      "phone": "phone",
      "postcode": "postcode",
      "recurrenceDates": Array [
        "03/08/2018",
        "04/08/2018",
      ],
      "stage": false,
      "startTime": "2018-08-02T00:00+00:00",
      "ticketingUrl": "ticketingUrl",
      "venueDetails": Array [
        "TBIuOS1mRo7nM",
        "BKX3JLP68i0k2e",
      ],
    },
    "id": "l503MHwD",
    "locale": "en-GB",
    "revision": 1,
  },
  Object {
    "contentType": "event",
    "fields": Object {
      "accessibilityDetails": "accessibilityDetails",
      "accessibilityOptions": Array [
        "FM6jre",
        "b8Y2Ioy1qEC",
        "0D98Lcd5il86y1sf4685Ba3s3RbTi4",
      ],
      "addressLine1": "addressLine1",
      "addressLine2": "addressLine2",
      "audience": Array [
        "X2T77u30HfGQ978Ce",
        "IsJqxf058AxsLcfAP7451huX7",
        "",
      ],
      "city": "city",
      "email": "email",
      "endTime": "2018-08-03T03:00+00:00",
      "eventCategories": Array [
        "Exhibition and Tours",
        "Community",
      ],
      "eventDescription": "eventDescription",
      "eventPriceHigh": 10,
      "eventPriceLow": 0,
      "eventsListPicture": Object {
        "sys": Object {
          "id": "7qByW",
        },
      },
      "individualEventPicture": Object {
        "sys": Object {
          "id": "pI1b5E4m4z",
        },
      },
      "location": Object {
        "lat": 0,
        "lon": 10,
      },
      "locationName": "locationName",
      "name": "name",
      "performances": Array [
        Object {
          "sys": Object {
            "id": "97hZrki0698g88ozhhp",
          },
        },
        Object {
          "sys": Object {
            "id": "48V92917",
          },
        },
        Object {
          "sys": Object {
            "id": "u6dtAeCrx2QtSFw7tlYU8Bt2Pgh",
          },
        },
        Object {
          "sys": Object {
            "id": "Qo9LdGIwhTV3VTYF5U5E1g4",
          },
        },
      ],
      "phone": "phone",
      "postcode": "postcode",
      "recurrenceDates": Array [
        "02/08/2018",
        "03/08/2018",
        "04/08/2018",
      ],
      "stage": false,
      "startTime": "2018-08-03T00:00+00:00",
      "ticketingUrl": "ticketingUrl",
      "venueDetails": Array [
        "TBIuOS1mRo7nM",
        "BKX3JLP68i0k2e",
      ],
    },
    "id": "l503MHwD-recurrence-03/08/2018",
    "locale": "en-GB",
    "revision": 1,
  },
  Object {
    "contentType": "event",
    "fields": Object {
      "accessibilityDetails": "accessibilityDetails",
      "accessibilityOptions": Array [
        "FM6jre",
        "b8Y2Ioy1qEC",
        "0D98Lcd5il86y1sf4685Ba3s3RbTi4",
      ],
      "addressLine1": "addressLine1",
      "addressLine2": "addressLine2",
      "audience": Array [
        "X2T77u30HfGQ978Ce",
        "IsJqxf058AxsLcfAP7451huX7",
        "",
      ],
      "city": "city",
      "email": "email",
      "endTime": "2018-08-04T03:00+00:00",
      "eventCategories": Array [
        "Exhibition and Tours",
        "Community",
      ],
      "eventDescription": "eventDescription",
      "eventPriceHigh": 10,
      "eventPriceLow": 0,
      "eventsListPicture": Object {
        "sys": Object {
          "id": "7qByW",
        },
      },
      "individualEventPicture": Object {
        "sys": Object {
          "id": "pI1b5E4m4z",
        },
      },
      "location": Object {
        "lat": 0,
        "lon": 10,
      },
      "locationName": "locationName",
      "name": "name",
      "performances": Array [
        Object {
          "sys": Object {
            "id": "97hZrki0698g88ozhhp",
          },
        },
        Object {
          "sys": Object {
            "id": "48V92917",
          },
        },
        Object {
          "sys": Object {
            "id": "u6dtAeCrx2QtSFw7tlYU8Bt2Pgh",
          },
        },
        Object {
          "sys": Object {
            "id": "Qo9LdGIwhTV3VTYF5U5E1g4",
          },
        },
      ],
      "phone": "phone",
      "postcode": "postcode",
      "recurrenceDates": Array [
        "02/08/2018",
        "03/08/2018",
        "04/08/2018",
      ],
      "stage": false,
      "startTime": "2018-08-04T00:00+00:00",
      "ticketingUrl": "ticketingUrl",
      "venueDetails": Array [
        "TBIuOS1mRo7nM",
        "BKX3JLP68i0k2e",
      ],
    },
    "id": "l503MHwD-recurrence-04/08/2018",
    "locale": "en-GB",
    "revision": 1,
  },
]
`;

exports[`Events reducer initialises with default state 1`] = `
Object {
  "amenities": Array [],
  "events": Array [],
  "featuredEvents": Array [],
  "headerBanners": Array [],
  "images": Object {},
  "loading": true,
  "noDataReceived": false,
  "paradeGroups": Array [],
  "performances": Object {},
  "refreshing": false,
  "sponsors": Array [],
}
`;
